src/python/pants/backend/go/module.py

```diff
@@ -0,0 +1,203 @@
+    import_path: str
+    minimum_go_version: Optional[str]
+    modules: FrozenOrderedSet[ModuleDescriptor]
+    digest: Digest
+
+
+@dataclass(frozen=True)
+class ResolveGoModuleRequest:
+    address: Address
+
+
+# Perform a minimal parsing of go.mod for the `module` and `go` directives. Full resolution of go.mod is left to
+# the go toolchain. This could also probably be replaced by a go shim to make use of:
+# https://pkg.go.dev/golang.org/x/mod/modfile
+def basic_parse_go_mod(raw_text: bytes) -> Tuple[Optional[str], Optional[str]]:
+    module_path = None
+    minimum_go_version = None
+    for line in raw_text.decode("utf-8").splitlines():
+        parts = line.strip().split()
+        if len(parts) >= 2:
+            if parts[0] == "module":
+                if module_path is not None:
+                    raise ValueError("Multiple `module` directives found in go.mod file.")
+                module_path = parts[1]
+            elif parts[0] == "go":
+                if minimum_go_version is not None:
+                    raise ValueError("Multiple `go` directives found in go.mod file.")
+                minimum_go_version = parts[1]
+    return module_path, minimum_go_version
+
+
+# Parse the output of `go mod download` into a list of module descriptors.
+def parse_module_descriptors(raw_json: bytes) -> List[ModuleDescriptor]:
+    module_descriptors = []
+    for raw_module_descriptor in ijson.items(raw_json, "", multiple_values=True):
+        module_descriptor = ModuleDescriptor(
+            import_path=raw_module_descriptor["Path"],
+            module_path=raw_module_descriptor["Path"],
+            module_version=raw_module_descriptor["Version"],
+        )
+        module_descriptors.append(module_descriptor)
+    return module_descriptors
+
+
+@rule
+async def resolve_go_module(
+    request: ResolveGoModuleRequest,
+    goroot: GoLangDistribution,
+    platform: Platform,
+    bash: BashBinary,
+) -> ResolvedGoModule:
+    downloaded_goroot = await Get(
+        DownloadedExternalTool,
+        ExternalToolRequest,
+        goroot.get_request(platform),
+    )
+
+    targets = await Get(UnexpandedTargets, Addresses([request.address]))
+    if not targets:
+        raise ValueError(f"Address `{request.address}` did not resolve to any targets.")
+    elif len(targets) > 1:
+        raise ValueError(f"Address `{request.address}` resolved to multiple targets.")
+    target = targets[0]
+
+    sources = await Get(SourceFiles, SourceFilesRequest([target.get(GoModuleSources)]))
+    flattened_sources_digest = await Get(
+        Digest, RemovePrefix(sources.snapshot.digest, request.address.spec_path)
+    )
+    flattened_sources_snapshot = await Get(Snapshot, Digest, flattened_sources_digest)
+    if (
+        len(flattened_sources_snapshot.files) not in (1, 2)
+        or "go.mod" not in flattened_sources_snapshot.files
+    ):
+        raise ValueError(f"Incomplete go_module sources: files={flattened_sources_snapshot.files}")
```

Comment:
This is better expressed w/ the Target API. Looks like we don't document this API yet, but:

https://github.com/pantsbuild/pants/blob/557113a2cf27d995f1ffc7cab0b13ddc37134bed/src/python/pants/engine/target.py#L1252

https://github.com/pantsbuild/pants/blob/557113a2cf27d995f1ffc7cab0b13ddc37134bed/src/python/pants/engine/target.py#L1260-L1269

(Make sure to call `super()` for `validate_resolved_files`.

Reply:
Ah, I was only aware of `expected_num_files`, which I considered using but didn't because it does not provide a way to ensure the filenames have specific names (i.e., `go.mod` and `go.sum`) so I just put the check in one place in this `if` statement. I will look into switching to `validate_resolved_files`.

Reply:
Fixed.